### PR TITLE
Update CMake fork of Libint2, c. 2019 -> c. 2022

### PR DESCRIPTION
**Pull request by loriab**
## Description

This corresponds to L2 at https://github.com/evaleev/libint/pull/233 .

## Todos

- [x] switches psi4 to use a Libint2 forked from upstream in 2022, not in 2019
- [x] changes a couple cmake knobs and changes the required components to EFV's preferred syntax (`g5` -> `eri_c4_d1_l5`)
- [x] avoided grid points discrepancies that I had hit by implementing Holger's suggestion.
- [x] I and others have reported needing `pybind11::headers` in extra places, so adding those
- [x] The defines `INT_NCART` and `INT_CARTINDEX` in psi https://github.com/psi4/psi4/blob/master/psi4/src/psi4/libmints/integral.h#L46-L64 are also in L2. They used to be `#define` in the current/2019 fork of L2, but now in L2 master, it’s an inline function. To avoid errors like the below, they're namespaced and inlined here now, too.

```
< namespace libint2 {
< inline int INT_NCART(int am) { return ((am + 2) * (am + 1)) >> 1; }
< }
< LIBINT_DEPRECATED("please use libint2::INT_NCART instead")
< inline int INT_NCART(int am) { return libint2::INT_NCART(am); }
---
> #define INT_NCART(am) ((((am)+2)*((am)+1))>>1)
```
With that situation (define in psi and inlined in L2), I get lots of
```
In file included from /psi/gits/hrw-pybind/psi4/src/psi4/libmints/integral.cc:28:
/psi/gits/libint2-efv/install-p01/include/libint2/cgshell_ordering.h:104:12: error: expected unqualified-id before 'int'
  104 | inline int INT_CARTINDEX(unsigned int am, int i, int j) {
      |            ^~~~~~~~~~~~~
```
I can fix the problem by renaming in psi to `INT_NCART_PSI` and `INT_CARTINDEX_PSI`. What's a cleaner solution than renaming? Should we inline? Or add to the `psi {}` namespace?

## Checklist
- [ ] ~Tests added for any new features~
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [ ] Ready for merge **SQUASH**

### CODE REVIEW

**Comment by andysim** [comment]
I think that renaming is perhaps the safest solution; without namespacing these kinds of macros it's possible to run into a situation where the two codes have the same macro name that returns a different result, which would cause chaos.  If you just want to get rid of the warnings (given that we know that L2 and Psi4 both have the same meaning of `INT_CART`) we could also just ensure that the L2 headers are included before Psi4 defines the macros; the `#if !defined(INT_CART)` will ensure we don't redefine the macro.

**Comment by andysim** [comment]
Oh, and to clarify by "namespacing" I'm referring to adding some kind of `PSI4` to the macro name, rather than adding the define within a namespace block.  To my knowledge the preprocessor wouldn't care about whether the definition occurs withing a C++ namespace or not.

File: external/upstream/libint2/CMakeLists.txt [hunk 1]
```diff
@@ -24,17 +25,16 @@
 # * If ever use a libint2 with AM>7, edit the `init_spherical_harmonics` line in libmints/integral.cc
 # * In selecting AM values, before `grep`ing lots of basis set files, consult the guide at end of this file.
 
-
 # CHOOSE! `make export`
 #find_package(Libint2 CONFIG)
 # CHOOSE! pure cmake
-find_package(Libint2 CONFIG COMPONENTS gss "e${MAX_AM_ERI}" g2 h2 eri3_e4 eri3_g3 eri3_h3 eri2_e4 eri2_g3 eri2_h3)
+find_package(Libint2 CONFIG COMPONENTS gss CXX_ho impure_sh "eri_c4_d0_l${MAX_AM_ERI}" eri_c4_d1_l2 eri_c4_d2_l2 eri_c3_d0_l4 eri_c3_d1_l3 eri_c3_d2_l3 eri_c2_d0_l4 eri_c2_d1_l3 eri_c2_d2_l3 onebody_d0_l4 onebody_d1_l3 onebody_d2_l3)
```
Comment (mbanck) [inline external/upstream/libint2/CMakeLists.txt:31]: :scream:

Comment (loriab) [reply external/upstream/libint2/CMakeLists.txt:31]: yeah, it's a little verbose, but that's a pretty minimal build AM-wise. it's going to be interesting to find a set of build conditions that all the OS QC packages can be content with.

**Review by maxscheurer — APPROVED** [review]

**Review by jturney — APPROVED** [review]
Looks good!

**Review by hokru — APPROVED** [review]

**Review by zachglick — APPROVED** [review]

**Pull request merged**